Fixed effectmanager.h to fill color

revision29 opened the pull request:
## Description
In the main branch this effect did not fill a solid color. Instead, it created a fire fill effect. This tweak fixes that. I am attempting this from a new branch created for this specific fix.

## Contributing requirements
<!-- Make sure your PR conforms to the requirements set out in CONTRIBUTING.md: -->

<!-- 
When ticking below boxes, please don't leave spaces between the 'x' and the square brackets, as that breaks the checkbox rendering in the PRs.
Right: [x]
Wrong: [x ]
-->
* [x] I read the contribution guidelines in [CONTRIBUTING.md](http://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/CONTRIBUTING.md).
* [x] I understand the BlinkenPerBit metric, and maximized it in this PR.
* [x] I selected `main` as the target branch.
* [x] All code herein is subjected to the license terms in [COPYING.txt](http://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/COPYING.txt).

rbergen commented:
For the record I'll repost the comment I originally made in #455: I think @davepl did this to show a "monochromatic" effect _that responds to audio input_ on configurations that have audio enabled, and might prefer things to stay that way.

I'll check with him and hold on this PR until we have clarity.

revision29 commented:
I can see the case for having a monochromatic effect that responds to sound. As written before I touched it, it did not produce a monochromatic effect. It actually created a fire fill effect. This method being triggered is supposed to fill a solid color and is called from the IR remote to fill a solid color. The IR remote iterates an array until it finds the appropriate color assigned to a button and sends it to this method. It should draw that color. I can imagine the fire fill effect is being triggered instead due to a custom project being implemented. Leaving the code as is completely breaks the functionality of drawing a single color to a strip. I will make a small tweak to my pull and use the color in the musical fire effect.

I don't want to be a major pain the the butt, but a method should do what is says it will do. Filling a strip with fire instead of a solid color is a little misleading. It's one of those things that causes noobs like me hours trying to diagnose problems that turn out to not be our fault.

revision29 commented:
Hopefully my fresh merge from the main branch didn't throw off this request.

I modified my code to reincorporate the musical fire effect when audio is enabled. I did not previously notice that effect was using the color received into the method. But, to keep it looking monochromatic I modified the custom fire effect. Instead of using black, white, yellow, and the color passed in, I replaced the yellow with a hue shifted version. It changes the hue by 10 which keeps it close enough to look monochromatic but different enough to not output the same color. That's the theory anyway.

After these changes: 
When audio is enabled, it creates a music reactive monochrome fill effect.
When audio is disabled, it draws the solid color on all pixels.

rbergen commented:
@revision29 Did you apply and push the update to your effectmanagercolorfix branch? I ask because it does not seem to have made it to this PR.

rbergen commented:
@revision29 Just making sure we're not waiting for each other: I'm waiting for the commit that restores the use of a (differently set up) musical fire effect when audio is enabled, to hit this PR. Without that I can't proceed with this (and that includes asking @davepl to have an opinion on your proposed change.)

Let me know if there's anything I can do to help with this.

rbergen commented:
> From an internal API perspective, we're likely to agree that a call to
void SetGlobalColor(CRGB color) {} resulting in flames violates the principle of least surprise.

I think it makes sense to check with Dave what the point of the method actually was, as I agree the implementation across projects/configurations is inconsistent. That is exactly what I intend(ed) to do anyway, but I still want to do that based on a "stable" alternative proposal, even if one could argue equally for and against that one. (For one, it illustrates that the current composition seems arbitrary as well.)

Personally, I also don't like the fact that EffectManager is mostly set up as an "effect agnostic arbiter" and then decides to pull in specific effects of its own accord.

> If the current effect wants to know that the GlobalColor changed (not unreasonable) and special case that to do weird things, we should enable that. 

I say we can simply do that by adding a `GetGlobalColor()` function, which any effect that wants to know what it is can pull in their `Draw()` function - which I think really is the only sensible place it could need it. It's also consistent with the general pattern for effects, which is "pull what you need when you need it". Similarly, we don't even tell effects that they are the current effect, instead it's a matter of "you better draw when you're told to".

Anyway, let's get some input on original intent - which I'd still prefer to do with the commit that's now missing - and move forward from there.

rbergen commented:
@robertlipe 

First off: I promise I read your whole reply, even though I'm not quote-replying to most of it. 

> Yes! It's that A student that decides to spray paint the gym at prom. :-)
> It tries so hard to be good and then it runs off and does the silly thing.

Still, it was done like this for a reason. I maintain that that reason should be the starting point for any potential decision to change it.

> Boo.

I think it's easy to forget what the core of NDS _really_ is: an "endless" sequence of `Draw()` invocations, indeed across various effects. Which is actually the real reason I'm so careful with spending bytes on "my own" JSON stuff - aside from the fact that RAM is still sparse in quite a few of the devices we support. The JSON stuff is supporting logic, it's very nice but not critical for the core function. This is actually a statement that puts _most_ of my own code contributions in perspective. Most of it is icing on the cake, if you will. The vast majority of resources should therefore be kept available to what is key, which is whatever is happening on the LEDs.

Based on that same principle of the drawing of effects being the core:

- We need to remember that in most cases, _all_ pixels attached to the controller need to be redrawn "at least 30 times per second", anyway. Which means that the cost of setting the LEDs will basically always far outweigh the costs of pulling in whatever facts the effect needs to be able to do so. What's also relevant here is that most "getters" are (automatically) inlined functions returning a class instance member variable. By the time the build toolchain is done with the code, that'll effectively be as cheap as reading a local variable in the vast majority of cases.
- Effects need to be ready to `Draw()` "at any time", which is why instantiating them properly, once, makes sense. There are quite a few reason why an effect may be started or stopped, like it reaching its "default run time" (which may be because the default run time is shortened by the user), the effect being specifically selected by a user (via "zapping" on the remote or activating it via the web UI/API), a temporary effect being activated, pixel data coming in via WiFi, etc. etc. Keeping track of all of these things and translating them to (partial) loads and unloads of effects "on the fly" is a massive undertaking that would consume resources I think we should reserve for the main function - `Draw()`. In fact, I'm convinced that some of what we would need to do would inevitably remove some of the smoothness of the visual experience we now get.
- There are _many, many_ things that an effect might decide are relevant for how it draws itself, but for any individual effect the vast majority of those won't be. So again, pulling in what an effect does need for whatever frame it needs to draw is relatively cheap. And when it isn't, the effect can (and should) indeed keep a "local" copy of the datum in question - like the weather effect does for properties that should trigger a load of weather data from the Network. 

Implementing some sort of pub/sub pattern for all state changes throughout the system that _may_ be relevant for whatever effect is running the LEDs, or may do so in a second, would:

- Introduce a massive complication to the project's programming model, at the "pub" side. Not just when it's initially implemented (it would need to cover a lot of what's already there), but from then on - any property we add anywhere should henceforth be pondered over whether it may be "`Draw()`-relevant" or not. If so, another pub/sub implementation would have to be added.
- Similarly, make the implementations of effects far more complicated. All of a sudden, effects need to figure out what it is they need to set themselves up to be notified about _and keep track of all those facts_. Now you're no longer talking about one `Draw()` function that gets what it needs to do its job, but also subscriber/listener functions that are then supposed to retain state.
- Cause a lot of concern (and CPU time, memory, ...) to be spent on setting up and sending out events that in many cases nobody cares about, _but theoretically might_.

In operating systems:

- Many possible changes need to be dealt with in _some_ way, and possibly by multiple entities (windows, controls, applications, ...) at the same time.
- The paradigm is "be kind to your neightbors, so only act when needed".

Then setting up event loops, sending around events, _and_ doing things like implementing default handlers, makes sense.  
Here at any one moment we have one effect, or even more accurately one `Draw()` function, that should be allowed to use all resources it can possibly (okay, maybe reasonably) be given. Everything that may "distract" from that should be approached with great care if not hesitation.

> I know that's a deep design choice that I sense is dear to you.

Actually, it's not dear to me. It wasn't my choice to begin with, and design choices are never "dear" to me. They make sense or they don't, which _can_ change over time. In this case I think it made sense when it was made and still does today, as I've just explained.

Back to the global color:

- If there is a way to apply the selected color in a (somewhat) persisted way, like it is now done for the HUB75, then I think `SetGlobalColor()` should do so.
- If there is no such way for some scenarios, I stand by my opinion that adding a getter that an effect can call if it's interested in whatever the global color is, is the thing to do.
- Adding a getter would not even require another "fact" to be remembered by `EffectManager` in this case, as it remembers the last selected color anyway - although the relevant class member variable's name would have to be changed, in my opinion.

revision29 commented:
@rbergen I guess we were waiting on one another. I thought for sure I pushed the change before stepping away for a few days. That provided to be false, but has now been remedied.  

In the mean time, the conversation took som interesting turns. For a future change, it would be nice to be able to persist the global color. I know the original intent was to have global color be a temporary effect triggered by a remote. Persisting the global color would be useful for accent lights in a room that a person might want to "set and forget." Definitely something to look into later.

revision29 commented on include/effectmanager.h:
```diff
@@ -342,13 +354,14 @@ class  EffectManager : public IJSONSerializable
             else
```
You would be correct. Not sure why the white was singled out to begin with. Will adjust when I go to correct typos.

revision29 commented on include/effectmanager.h:
```diff
@@ -329,6 +329,18 @@ class  EffectManager : public IJSONSerializable
+        
+        //This block shifts the hue of the received color to be used in the audio reactive monochromatic fire effect.
+        //Becaue of scope errors by the compiler, the color conversion happens here.
+        CHSV colorVarientHSV = rgb2hsv_approximate(color);
```
I will be glad to fix the formatting. I really just wanted to explain my work for the pull request and can remove the comments completely. Although someone might read the code later and wonder why that code block is in the wrong place. 

A varient is when a person intends to write "variant" but, is bad at spelling and writes the same misspelling over and over. This is an embarrassing mistake. Will fix.

As to the placement of the color shift block: It bugs me that I had to put it there. It made no sense as to why it kept failing to compile. I should have captured the error to be more precise. So, just now, I put the code where it makes sense in order to se what the error was. It compiled just fine. I recently rebooted the computer. Maybe that did the trick. No offense meant towards gcc. The problem probably lies in the complex relationship between VS Code, platform IO, whatever plugins are installed, and the fact that my Mac probably is not happy with all of the MSSoft software installed.

Wrapping around the hue: I will trust you if you say it will safely wrap around. I prefer to test it before I full send, but I don't have the correct hardware. I picked up some microphones to wire to my board. But, since I'm using a Heltec Wi-FI Kit V3 and not the m5stick, there are many, many errors thrown at compile time when enabling audio and setting the input pin. Anyway, the worst that will happen without warping is that if a color has a hue of 255, the variant will just be the same color. Truly monochromatic.

Fresh commit incoming.

rbergen commented:
So, I've had my chat with Dave, and the intent was for whatever the current effect is to adopt the hue of the color button that was pressed on the remote. Which `SetGlobalColor()` does for the HUB75 scenario to some extent at least, but for none of the others. Which may be because it could well require each "WS2812" effect to adopt the "global color" in a way that suits it. Adding that to all effects we have is a fair amount of work.

If we stick to the original intent then forcing the effect to anything else than the current effect is a bit of a cop-out, and the choice of effect we change to if we still choose to do so is effectively arbitrary, to a point.

If I'm brutally honest then to me personally, the choice of effect triggered by the color buttons is a case of half a dozen of this vs. six of the other. That in turn means I'm happy to take a majority vote on merging this - provided it builds and runs, of course.

About that, there is a "build_all.py" script in the tools directory (do note the run instructions at the top of the file) to build all configurations locally. Alternatively or additionally, one can enable GitHub Actions in one's fork and have GitHub/Azure do the job at every push - the benefit there being that it will run the builds in parallel, meaning it finishes building all 34 configurations in about 10 minutes.

rbergen commented:
> If you force this to build, you'll find it won't build since you still have "variEnt" here.https://pine64.com/product/pinenut-model01s-wifi-ble5-module/

@robertlipe Although you probably intended to paste a different link than the one you did, it's correct that CI on "WS2812 audio-enabled configurations that aren't Spectrum" currently fails for this reason.

revision29 commented:
@rbergen Knowing Dave's intent helps to clarify the code. My only counter is that he might have intended to do one thing, but ended up doing another. Setting a global color that an effect might use is a great idea. I have not looked closely, but don't think any effects are coded to work that. To muddy the waters even more, the original code triggered a fire fill effect instead of interacting with whatever effect was currently active. So, I am seeing a disconnect between the intent and the actual code. That more than likely results to the exact use case he was making the code for.

I am thinking I might go ahead and cancel the pull request and reach the my desired result in a different way. Before I do, I would like to get some help in finding out why the scope error is popping up in some builds.

Based on Dave's explanation, here is what I think should happen when the set global color method is invoked:
1) Set the global color in effect manager.
2) Not draw anything.
3) Detect what current effect is running.
4) Restart the current effect.
5) If it is coded to do so, the current effect reads the global fill color and draws itself accordingly. If the global fill color is black or all 0s in RGB the effect draws itself as normal.

Of course this dos not make sense in every effect, but certainly in some.

I would not mind trying to tweak the current strip effects to interact with the global fill color in effect manager to realize Dave's intent.  I think the original intention should stand. It would be a long term improvement I could hack away at AFTER I get done reworking the remote control code. But, I am not touching HUB75 effects unless a HUB75 matrix mysteriously shows up in my mailbox (my wife is already noticing an abundance of Amazon packages as of late:) ). I would, however, rework the ColorFillEffect rather soon to take advantage of this.

But, for the sake of my own sanity and knowledge, why is the scope error showing up?

revision29 commented:
I have been trying to figure out how to not use SetGlobalColor to set all pixels to the same color. It is possible to use the strip effect "ColorFillEffect". Seems simple. Don't know why I didn't think about it before.  I would be able to create a new instance of that effect defined with the color I want it to draw and send it to effect manager. But, there is no way to send a temporary effect to EffectManager. 

There is another problem. Effects cannot access the global color. The variable in effects manager that stores that color (lastManualColor) is not accessible outside of the class / object itself. If we want effects to take advantage of the global color, they will need access to the lastManualColor variable.

As things stand now, there is no way to press a color button on a remote and have a led strip show that color, unless we go with my changes to the SetGlobalColor method. But, if we leave as is, the intent to have effects utilize the global color won't work because effects don't have access to the global color variable. This has left me frustrated. Neither case works unless a person is using the exact custom fire effect specified in code.

Would it be acceptable to modify effect manger to:
1) Create a new method for setting a temporary effect?
2) Create a new method for returning the current global color?
3) SetGlobalColor draws nothing for regular strips when a) audio is off or b) audio is on but spectrum is off?

rbergen commented:
> I have been trying to figure out how to not use SetGlobalColor to set all pixels to the same color. It is possible to use the strip effect "ColorFillEffect". Seems simple. Don't know why I didn't think about it before. I would be able to create a new instance of that effect defined with the color I want it to draw and send it to effect manager. But, there is no way to send a temporary effect to EffectManager.

Actually, the current code in `SetGlobalColor()` sets the temporary effect for all cases except HUB75. It does so by setting `_tempEffect`. It is true that there isn't a (public) member function to do the same from outside of `EffectManager()`. Of course it is possible to add that.

The thing I'm not entirely getting though, is what we would gain exactly from not setting the temporary effect within `SetGlobalColor()`, but then doing exactly the same from "outside" of `EffectManager` - if we really want to make `SetGlobalColor()` determine the hue of the current effect, that is.

> There is another problem. Effects cannot access the global color. The variable in effects manager that stores that color (lastManualColor) is not accessible outside of the class / object itself. If we want effects to take advantage of the global color, they will need access to the lastManualColor variable.

Adding a getter function that returns `lastManualColor` (which I think we should then rename to `_globalColor`, but I digress a little) is basically what I've been suggesting we do in a few previous comments, if we want effects to be able to retrieve and apply the global color (or actually, hue) to themselves.

> Would it be acceptable to modify effect manger to:
> 
> 1. Create a new method for setting a temporary effect?

In principle yes, but I'm still not sure what purpose this would serve.

> 2. Create a new method for returning the current global color?

Yes.

> 3. SetGlobalColor draws nothing for regular strips when a) audio is off or b) audio is on but spectrum is off?

Again, how would this improve the current overall behavior?

Have we actually already established that FastLED (or another supporting library) does not allow the hue of all LEDs (preferably on all channels) to be changed at once, in a sort of "overlay" fashion? If that would be possible then we could quite easily implement the originally intended behavior.  
Or, if that's not possible, implement modification 2 to EffectManager and have `LEDStripGFX::PostProcessFrame()` "overlay" the selected hue?  
We'd probably have to think how to "reset" the selected color to "nothing" upon a particular remote control button push and/or when the current effect changes, but those things are definitely solvable.

davepl commented:
I guess I should give a little brain dump on the history and the intent!

It started as a way to get a solid one color spectrum analyzer.  So when you hit a remote color button, you set the global color.
If a global color is set, the spectrum will use it.  
If you hit a SECOND color button, it then makes a palette fade across the two colors and uses that.  It calls this “Palette Paused” mode.

Now here’s the rub.  Here’s what the effects should ALL do:
- Draw as normal.  If an accent color or hue makes sense for that effect, it should follow the global current palette
- If a global color is set, it should use the global color
	- Not all effects can/should… like weather might only change an accent color, or nothing at all

This is non-trivial to do across all effects, so I’d suggest starting with the biggest bang for your buck by coloring Fireplace and other effects

This global color is used differently by different effects.  Like on the atomic fire lamp, a color button will set all the spokes to that color.

So there are really two things we should do that we’re not consistent with:
- If an effect can use an accent color, use the current rotating palette, unless a Color button (paused palette) is set, then it should use that color

While it’d be nice if all effects did it consistently, there’s no doubt a few like fireplace or ghost wave where it makes the most sense to start there…

- Dave

> On Oct 14, 2023, at 12:33 PM, Joe Schneider ***@***.***> wrote:
> 
> 
> I have been trying to figure out how to not use SetGlobalColor to set all pixels to the same color. It is possible to use the strip effect "ColorFillEffect". Seems simple. Don't know why I didn't think about it before. I would be able to create a new instance of that effect defined with the color I want it to draw and send it to effect manager. But, there is no way to send a temporary effect to EffectManager.
> 
> There is another problem. Effects cannot access the global color. The variable in effects manager that stores that color (lastManualColor) is not accessible outside of the class / object itself. If we want effects to take advantage of the global color, they will need access to the lastManualColor variable.
> 
> As things stand now, there is no way to press a color button on a remote and have a led strip show that color, unless we go with my changes to the SetGlobalColor method. But, if we leave as is, the intent to have effects utilize the global color won't work because effects don't have access to the global color variable. This has left me frustrated. Neither case works unless a person is using the exact custom fire effect specified in code.
> 
> Would it be acceptable to modify effect manger to:
> 
> Create a new method for setting a temporary effect?
> Create a new method for returning the current global color?
> SetGlobalColor draws nothing for regular strips when a) audio is off or b) audio is on but spectrum is off?
> —
> Reply to this email directly, view it on GitHub <https://github.com/PlummersSoftwareLLC/NightDriverStrip/pull/456#issuecomment-1763157579>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA4HCF3LL5WYVOOOTUVMUJ3X7LSIJANCNFSM6AAAAAA5W7RZNU>.
> You are receiving this because you were mentioned.
>

revision29 commented:
@rbergen 
> Adding a getter function that returns lastManualColor (which I think we should then rename to _globalColor, but I digress a little) is basically what I've been suggesting we do in a few previous comments, if we want effects to be able to retrieve and apply the global color (or actually, hue) to themselves.

Based on Dave's clarification, we will need two global color variables: _previousGlobalColor and _globalColor. I appreciate your patience with me in all of this. I feel like a bull in a china shop. It takes some mental and verbal (in this case, text) processing for things to click in my mind. Anyway, should a getter method be created or just set the two variables to public and be able to access them directly?  My opinion is the second would be a little more efficient.

> Create a new method for setting a temporary effect?
> In principle yes, but I'm still not sure what purpose this would serve.

The purpose is that I want to be able to trigger a custom effect with the press of a button. The 42 key remote has 6 DIY buttons. I'd like to allocate those to a trigger specific effect. This is to allow greater flexibility / customization in the system. In a project I am working on for my office, I have 5 or so similar effects with different color palettes loaded on boot. I can rotate through the list with a next button on my remote. But, I also want to be able to trigger a specific custom effect outside of that loop. It's a whole lot easier than adding a disabled effect on boot, remembering what index it is in the effect array, and hard coding that in my DIY button functionality.


> SetGlobalColor draws nothing for regular strips when a) audio is off or b) audio is on but spectrum is off?
> Again, how would this improve the current overall behavior?

SetGlobalColor forces a specific effect. That is not how Dave has described how the method should work. It should only set the global color, demote the current global color to secondary, and then trigger the current effect to refresh using the effect manager Update() method. If the effect is coded properly (based on future changes to those effects) it will read the global color and behave accordingly, as we have been discussing. Now that I understand Dave's intention more clearly, I am rally excited to make it work like that. For example, I can load ColorFillEffect as one of my default effects and then use the SetGlobalColor to change it instead of triggering a temporary effect for every color button (my current way of getting a solid fill color).

> Have we actually already established that FastLED (or another supporting library) does not allow the hue of all LEDs (preferably on all channels) to be changed at once, in a sort of "overlay" fashion? If that would be possible then we could quite easily implement the originally intended behavior.

From my hours of experimenting, as far as I can tell, the only way to overlay is to touch each and very pixel that we want to change. That is exactly how brightness is being applied. From what I have gathered, that is the only way. I think any other solution would be overly complicated. The effect itself should be able to implement a change in color. This is a rabbit hole I am not willing to dive down.

Anyway, I can get to work on getting things moving in the direction I think the discussion is moving towards. 

But, before I do, can you provide any insight as to why the change I proposed in this pull request is failing to build in some instances? The scope error makes no sense because the variable it is throwing an error about is being created in the same scope both in the script logic and within compiler #if / #else logic.

revision29 commented:
@davepl thank you for the clarification. I made too many bad assumptions when I was reading through the code and trying to make changes to fit my use cases. One last question for clarification: in the SetGlobalColor method, you had it trigger a temporary fire effect based on the global color when some conditions were met. Are you saying that such an effect should not be triggered, but that an effect currently running or in the effect manage queue should read that color and draw accordingly? In other words, it's okay to remove the custom fire effect and just have the method change the global color variable and reload the current running effect? At the same time, I would be glad to change MusicalPaletteFire and PaletteFlameEffect to make use of the global color since you are using those in SetGlobalColor.

revision29 commented:
So, with some more experimentation, I found some limitations with effects directly interacting with EffectManager and any stored properties, i.e. _globalColor. Any effects loaded from misceffects.h and fire effect.h cannot access g_ptrSystem->EffectManager(). This is because those two files are loaded before the EffectManger is initialized. Trying to load those files later in the sequence predictably breaks things badly. So, I guess the global color is going to have to become a global variable. I didn't want it to come to this, but here we are. That is, unless someone else has a better idea.

Anyway, I forgot to submit this comment and went off experimenting. I put the global color in g_Values and changed the color fill effect to use that global color if the color is not null or set to black. I put the color fill effect in the set of effects loaded at boot. I changed the color with my remote, but since the color effect was not active nothing changed. That was expected. Then I used the button I designated for NEXT_EFFECT to move through the effects loaded in effect manager. When I landed on the color fill effect it showed the color I set with the remote. Success. The concept works.

davepl commented:
One general sentiment I wanted to throw out is that nothing is really sacred in the code!  

My personal level of attachment to a piece of code is probably linearly correlated with how much time I put into thinking about, writing, debugging, and using a particular piece of code.  In case where that’s a small number, feel no compunction about changing things.

Conversely, the round robing queuing system for wifi packets isn’t ripe for a rewrite, because it works, is well tested, and reliable.  

Feel free to do whatever is right for (a) the user experience first, and (b) the codebase’s readability and performance second.  

If I could summarize my philosophy in bullet points, it would be:

- No sacred cows; 
- Don’t change stuff just for the sake of changing it in your own image or I will walk through your wall and revert the commit
- The “right thing” is what you’d ideal want it to do if you* had one on your bookshelf in your room

(*) I have autism, so that means what I would want it to do, as I assume everyone things the same as me.  But if you can explain why a behavior different from my assumtptions is better for users, then I’m usually open!

Offhand, perhaps we should expose GetGlobalColor() and GetGlobalPalette().  They do what I described (one color or a range of colors) and the implementation of how to do that is left to whoever writes the code!  Fewer global variables is better I think.  That’s something I’d like to further reduce longterm…

I think there are only 5 effect buttons - the top right one is the WHITE button, or that’s how mine are labelled.

How about a simple table mapping IR id to effect?  If a button has an effect entry in the table, then that button launches that effect.   Could be as simple as effect index in the main effect table, so a table of int tuples.  Then different configs could do different things… because the Atomic Fire Lamp has different needs than the Mesermizer or the Fire Fans.  Just thinking out loud.

Cheers,
Dave


> On Oct 15, 2023, at 12:17 PM, Joe Schneider ***@***.***> wrote:
> 
> 
> @rbergen <https://github.com/rbergen>
> Adding a getter function that returns lastManualColor (which I think we should then rename to _globalColor, but I digress a little) is basically what I've been suggesting we do in a few previous comments, if we want effects to be able to retrieve and apply the global color (or actually, hue) to themselves.
> 
> Based on Dave's clarification, we will need two global color variables: _previousGlobalColor and _globalColor. I appreciate your patience with me in all of this. I feel like a bull in a china shop. It takes some mental and verbal (in this case, text) processing for things to click in my mind. Anyway, should a getter method be created or just set the two variables to public and be able to access them directly? My opinion is the second would be a little more efficient.
> 
> Create a new method for setting a temporary effect?
> In principle yes, but I'm still not sure what purpose this would serve.
> 
> The purpose is that I want to be able to trigger a custom effect with the press of a button. The 42 key remote has 6 DIY buttons. I'd like to allocate those to a trigger specific effect. This is to allow greater flexibility / customization in the system. In a project I am working on for my office, I have 5 or so similar effects with different color palettes loaded on boot. I can rotate through the list with a next button on my remote. But, I also want to be able to trigger a specific custom effect outside of that loop. It's a whole lot easier than adding a disabled effect on boot, remembering what index it is in the effect array, and hard coding that in my DIY button functionality.
> 
> SetGlobalColor draws nothing for regular strips when a) audio is off or b) audio is on but spectrum is off?
> Again, how would this improve the current overall behavior?
> 
> SetGlobalColor forces a specific effect. That is not how Dave has described how the method should work. It should only set the global color, demote the current global color to secondary, and then trigger the current effect to refresh using the effect manager Update() method. If the effect is coded properly (based on future changes to those effects) it will read the global color and behave accordingly, as we have been discussing. Now that I understand Dave's intention more clearly, I am rally excited to make it work like that. For example, I can load ColorFillEffect as one of my default effects and then use the SetGlobalColor to change it instead of triggering a temporary effect for every color button (my current way of getting a solid fill color).
> 
> Have we actually already established that FastLED (or another supporting library) does not allow the hue of all LEDs (preferably on all channels) to be changed at once, in a sort of "overlay" fashion? If that would be possible then we could quite easily implement the originally intended behavior.
> 
> From my hours of experimenting, as far as I can tell, the only way to overlay is to touch each and very pixel that we want to change. That is exactly how brightness is being applied. From what I have gathered, that is the only way. I think any other solution would be overly complicated. The effect itself should be able to implement a change in color. This is a rabbit hole I am not willing to dive down.
> 
> Anyway, I can get to work on getting things moving in the direction I think the discussion is moving towards.
> 
> But, before I do, can you provide any insight as to why the change I proposed in this pull request is failing to build in some instances? The scope error makes no sense because the variable it is throwing an error about is being created in the same scope both in the script logic and within compiler #if / #else logic.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/PlummersSoftwareLLC/NightDriverStrip/pull/456#issuecomment-1763479794>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA4HCF6OHPXWTT76QTNMM4TX7QZDPANCNFSM6AAAAAA5W7RZNU>.
> You are receiving this because you were mentioned.
>

revision29 commented:
@davepl 

This is my second draft of this post. I thought I learned my lesson the last time I nuked a long comment by pressing the "go back in history" keyboard command in my browser. This is the shorter version. TLDWTTA (too long, don't want to type again).

> Conversely, the round robing queuing system for wifi packets isn’t ripe for a rewrite, because it works, is well tested, and reliable.

I am here to mess with effects and IR remote functionality. There are zero plans to go near that code.

> Don’t change stuff just for the sake of changing it in your own image or I will walk through your wall and revert the commit

This "threat" has Freddy Kruger vibes.


> The “right thing” is what you’d ideal want it to do if you* had one on your bookshelf in your room.

That is the exact approach I am taking. I am here because your videos inspired me to learn how to mess with LEDs and make some decorations. What I am making will be hanging on my wall, sitting on my bookshelf, lining my desk, etc. I am trying to make improvements and help increase usability based on real world implemations. I actually just finshed soldering up my first piece which will be mounted on my office wall. 

> GetGlobalColor() and GetGlobalPalette().

I love the idea, but this is not currently possible. Some effects are included before the effect manager is initialized. g_ptrSystem->EffectManager() is not accessible by those effects.  I tested something like this last night and it was not workable, at least for me. This is a web I don't want to try and untangle for a long time. For the time being, I vote to put the global color as a global variable knowing that we will work to change that later.

> How about a simple table mapping IR id to effect? 

I am currently reworking the IR control system. I keep getting distracted by other things along the way. My solution has the following components:
1) ENUM with most of the possible actions a button might do.
2) A button class that holds three variables: label; enum action; arguments. The arguments variable is to pass along any variables to the action. For example, my remote has 20 color buttons. I am not making 20 enum entries. Instead, I define 20 color buttons that pass along a hex rgb code. 
3) A std::map of button objects. Each key is the IR code for the button. The value is the button object.
4) When the remote receives the code, it seaches the map for a matching key. This is improved over my previous attempt using std::vector.
5) We run a switch against the action stored in the button object to trigger the desired effect.

I am trying to maximize customizability in a way that is not overly complex. I will make sure to comment the code to help people understand how to customize their remote. Samples of the code. I can guarantee that the web interface will mess up the formatting, but you will be able to read it well enough.

Example of the buttons std::map:
> //Row 1
> buttons.emplace(0xFF3AC5, RemoteButton("Brightness Up",BRIGHTNESS_UP));
> buttons.emplace(0xFFBA45, RemoteButton("Brightness Down",BRIGHTNESS_DOWN));
> buttons.emplace(0xFF827D, RemoteButton("Next Effect",NEXT_EFFECT));
> buttons.emplace(0xFF02FD, RemoteButton("Power Off",POWER_OFF));

> //Row 2
> buttons.emplace(0xFF1AE5, RemoteButton ("Full Red",FILL_COLOR, "FF0000"));
> buttons.emplace(0xFF9A65, RemoteButton ("Full Green",FILL_COLOR, "00FF00"));
> buttons.emplace(0xFFA25D, RemoteButton ("Full Blue",FILL_COLOR, "0000FF"));
> buttons.emplace(0xFF22DD, RemoteButton ("Full White",FILL_COLOR, "AAAAAA")); 

Example from the switch in remotecontrol.cpp
> 				case BRIGHTNESS_UP:
>                deviceConfig.SetBrightness((int)deviceConfig.GetBrightness() + BRIGHTNESS_STEP);
>             break;
>             case BRIGHTNESS_DOWN:
>                 deviceConfig.SetBrightness((int)deviceConfig.GetBrightness() - BRIGHTNESS_STEP);
>             break;
>             case POWER_OFF:
>                 effectManager.ClearTemporaryStripEffect();
>                 effectManager.ClearRemoteColor();
>             break;
>             case NEXT_EFFECT: 
>                 effectManager.NextEffect();
>             break;
>             case FILL_COLOR:
>             {
>                 lastManualColor = hexToCrgb(thisButton.actionArgs); // lastManualColor is saved locally in the remote control
>                 effectManager.SetGlobalColor(lastManualColor);
>             }
>             break;

rbergen commented:
@revision29 It's a long one, because I'm responding to a few comments in one go.

> Based on Dave's clarification, we will need two global color variables: _previousGlobalColor and _globalColor. 

Ok, so if we need two we add two. :)

> I appreciate your patience with me in all of this. I feel like a bull in a china shop. 

Don't worry, you're not that bull until you actually break stuff.

> Anyway, should a getter method be created or just set the two variables to public and be able to access them directly?

From a principle of encapsulation, in classes getter and setter functions should generally be used, certainly if anything needs to happen when the value is set. Which is certainly the case _now_, where all the logic in `SetGlobalColor()` is triggered. 

> My opinion is the second would be a little more efficient.

It won't make a meaningful difference. All functions are implicitly inlined, so the build toolchain will make the getter effectively a direct pull of the returned variable anyway.

> The purpose is that I want to be able to trigger a custom effect with the press of a button. 

Fair enough, seems like a good enough reason to me.

>>> SetGlobalColor draws nothing for regular strips when a) audio is off or b) audio is on but spectrum is off?

>> Again, how would this improve the current overall behavior?

> SetGlobalColor forces a specific effect. That is not how Dave has described how the method should work. 

No, it isn't. My point was that the new behavior needs to be in place before we pull out the old one, otherwise it's a regression instead of an improvement. If your intent is to also implement the new behavior in the various effects, then I'm all for doing this.

> But, before I do, can you provide any insight as to why the change I proposed in this pull request is failing to build in some instances?

It's a simple typo. In line 346 you declare the variable as `colorVarientHSV` (variEnt with an E), and in lines 347 and 350 you refer to it as `colorVariantHSV` (variAnt with an A). Which is what the compiler hints at with the last line of this log snippet: 

```log
In file included from include/globals.h:1547,
                 from src/colordata.cpp:32:
include/effectmanager.h: In member function 'void EffectManager::SetGlobalColor(CRGB)':
include/effectmanager.h:347:21: error: 'colorVariantHSV' was not declared in this scope
                     colorVariantHSV.hue += 10; // A hue-shifted version of the received color to build the monochromatic effect.
                     ^~~~~~~~~~~~~~~
include/effectmanager.h:347:21: note: suggested alternative: 'colorVarientHSV'
```

> Any effects loaded from misceffects.h and fire effect.h cannot access g_ptrSystem->EffectManager(). This is because those two files are loaded before the EffectManger is initialized. 

In the "regular flow" **no** effect is ever started or even initialized before the `EffectManager` is initialized. The order in which files are included is irrelevant at runtime, all that matters is the order in which code is executed. 
For completeness: the order in which files are included _is_ relevant at _build_ time, but that is something that can be fixed if it's currently broken.

To get started with this, please explain _how_ you concluded `g_ptrSystem->EffectManager()` can't be accessed?

> g_ptrSystem->EffectManager() is not accessible by those effects. I tested something like this last night and it was not workable, at least for me. This is a web I don't want to try and untangle for a long time. For the time being, I vote to put the global color as a global variable knowing that we will work to change that later.

I vote against. Let's figure this out now, because I've had very, very many experiences with "temporary" solutions sticking around forever. As you guessed, those usually weren't positive.  

There are already 2 effect header files that use `g_ptrSystem->EffectManager()` in their `Draw()` functions, so I really don't understand why others wouldn't be able to, or couldn't be made to.  
Just off the top of my head, could it be that there is an `#include "systemcontainer.h"` line missing in the `#include` sections at the top of the .h files in question? 

For the record: one reason I'm being so persistent about this is that if we want to make the global color (or colors, or color/palette combo, or...) persistent across reboots, they will have to end up in `DeviceConfig` at some point. That's also referenced through `g_ptrSystem`. So, if we have an issue there it's just better to get to the bottom of it early.

rbergen commented:
@revision29 It just occurred to me you may be running into a circular include problem, because misceffects.h and fireeffect.h are pulled into effectmanager.h. If you're changing the effects to take care of their own implementation of the global color(s), then there's no need for them to be instantiated in `SetGlobalColor()` any more. And if that's no longer necessary, the respective effect headers don't need to be pulled into effectmanager.h anymore either. So, including systemcontainer.h into misceffects.h and fireeffect.h and removing the includes of the latter two files from effectmanager.h should fix whatever it is you're now tripping over.

If what I'm suggesting is not possible for any reason, we can move part of the relevant effects' implementations to .cpp files. But let's not go there yet.

rbergen commented:
@revision29 I was wondering if you are still working on this PR, or do you intend to open another one in due course?

revision29 commented:
@rbergen I intend to close this and start another. But first, I was hoping for some insight about the scope error that occurs in some builds. The code that it complains about is within certain preprocessor conditions. If those conditions are not met, the cod should not even be sent to the compiler to generate any errors. If the conditions are met, the variables are created and accessed within the same scope. Can you explain in more detail what is happening and why? 

But yes, I found another way to the general concept of making all LEDs th same color.

rbergen commented:
@revision29 I actually explained this in this earlier comment: https://github.com/PlummersSoftwareLLC/NightDriverStrip/pull/456#issuecomment-1763883755, specifically the part where I refer to line 346.

It's a typo you made in the declaration of one of the variables. The "scope" thing is just a case of compiler semantics in the error it generates when it can't find an identifier (in this case variable) you're referring to.

revision29 commented:
@rbergen for some reason I stopped getting GitHub notifications and never saw the comment. After I made the bug report earlier all of a sudden flood of emails started to arrive.

I looked at those lines of code over and over and never noticed the last typo. Now the scope error makes total sense. I need to stop coding late at night, but that’s when I have the free time. 

Back to the previous discussion about the global color. In my experimentation, I did run into the circular inclusion problem as I was trying to play around with the order in which the header files were being included. I quickly backed off that. I reduced the number of errors by not creating an effect when the global color was set, but I still could not change the loading sequence without it failing to build. I need to dive a little deeper and experiment some more. I will have time for that on Saturday.

rbergen commented:
> @rbergen for some reason I stopped getting GitHub notifications and never saw the comment. After I made the bug report earlier all of a sudden flood of emails started to arrive.

I am just guessing, but it could be that GitHub pauses sending them if you don't interact with their platform for too long.

I've spent a little more time thinking about the circular include problem, and I think it's actually quite easy to fix. If you scroll to the bottom of effectmanager.cpp (i.e. **not** the header), you'll see that there are two helper functions there called `GetSpectrumAnalyzer()`, one of which is actually called by `EffectManager::SetGlobalColor()`. `GetSpectrumAnalyzer()` depends on spectrumeffects.h, which is almost certainly why things were set up this way.

What we can do is move the _definition_ of `EffectManager::SetGlobalColor()` (i.e. the actual logic) to effectmanager.cpp, leaving only its _declaration_ in effectmanager.h. If we do that, we can also move the effect headers it depends on (misceffects.h and fireeffect.h) to effectmanager.cpp. That should short out the circular dependency at a stroke. I'd put the function definition and the effect headers it needs just below the `GetSpectrumAnalyzer()`s, so that you can also remove the declarations for `GetSpectrumAnalyzer()` from the top of effectmanager.h.

rbergen commented:
@revision29 PR #474 that I have just opened actually implements the approach in my previous comment.

rbergen commented:
Closing after addressing the scope error [in this and later comments](https://github.com/PlummersSoftwareLLC/NightDriverStrip/pull/456#issuecomment-1763883755).
